File: 0971-flip-binary-tree-to-match-preorder-traversal/0971-flip-binary-tree-to-match-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int i=0;
    bool dfs(TreeNode* node,vector<int>& v,vector<int>& ans) {
        if(!node) return true;
        if(node->val != v[i++]) return false;
        if(node->left && node->left->val != v[i]) {
            ans.push_back(node->val);
            return dfs(node->right,v,ans) && dfs(node->left,v,ans);
        }
        return dfs(node->left,v,ans) && dfs(node->right,v,ans);
    }
    vector<int> flipMatchVoyage(TreeNode* root, vector<int>& voyage) {
        vector<int> ans;
        return dfs(root,voyage,ans) ? ans : vector<int>{-1};
    }
};

File: 0971-flip-binary-tree-to-match-preorder-traversal/0971-flip-binary-tree-to-match-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int i=0;
    public boolean dfs(TreeNode node,int[] arr,ArrayList<Integer> ans) {
        if(node == null) return true; //is equal
        if(node.val!= arr[i++]) return false; //is not equal set up i for child 
        //now i for child is set up and we can compare i with child.
        if(node.left!= null && node.left.val != arr[i]) {
            ans.add(node.val);
            return dfs(node.right,arr,ans) && dfs(node.left,arr,ans);
        }
        
        return dfs(node.left,arr,ans) && dfs(node.right,arr,ans);
    }
    public List<Integer> flipMatchVoyage(TreeNode root, int[] voyage) {
        ArrayList<Integer> ans = new ArrayList<>();
        return dfs(root,voyage,ans) ? ans : Arrays.asList(-1);
    }
}